src/csv/include/csv/parsing.hpp

```diff
@@ -36,13 +37,53 @@ class CsvReadException : public utils::BasicException {
   using utils::BasicException::BasicException;
 };
 
+class FileCsvSource {
+ public:
+  explicit FileCsvSource(std::filesystem::path path);
+  std::istream &GetStream();
+
+ private:
+  std::filesystem::path path_;
+  std::ifstream stream_;
+};
+
+class StreamCsvSource {
+ public:
+  StreamCsvSource(std::stringstream stream) : stream_{std::move(stream)} {}
+  std::istream &GetStream() { return stream_; }
+
+ private:
+  std::stringstream stream_;
+};
+
+class UrlCsvSource : public StreamCsvSource {
+ public:
+  UrlCsvSource(char const *url);
+};
+
+class CsvSource {
+ public:
+  static auto Create(utils::pmr::string const &csv_location) -> CsvSource;
+  CsvSource(FileCsvSource source) : source_{std::move(source)} {}
+  CsvSource(StreamCsvSource source) : source_{std::move(source)} {}
+  CsvSource(UrlCsvSource source) : source_{std::move(source)} {}
+  std::istream &GetStream();
+
+ private:
+  std::variant<FileCsvSource, UrlCsvSource, StreamCsvSource> source_;
+};
+
 class Reader {
  public:
   struct Config {
     Config() = default;
     Config(const bool with_header, const bool ignore_bad, std::optional<utils::pmr::string> delim,
            std::optional<utils::pmr::string> qt)
-        : with_header(with_header), ignore_bad(ignore_bad), delimiter(std::move(delim)), quote(std::move(qt)) {}
+        : with_header(with_header), ignore_bad(ignore_bad), delimiter(std::move(delim)), quote(std::move(qt)) {
+      // delimiter + quote can not be empty
+      if (delimiter && delimiter->empty()) delimiter.reset();
+      if (quote && quote->empty()) quote.reset();
+    }
```

Comment:
so what if they are empty? is it okay if we `spdlog::info` the resolution of this resetting behaviour

Reply:
Empty string opened up vulnerability issues, as highlighted by the fuzzer. 

Chose to fix here, rather than later in the code, as its actually a logical error. What does an empty delimiter or empty quote actually mean?